presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java

```diff
@@ -158,4 +165,87 @@ private static boolean hasSubscripts(Optional<Subfield> subfield)
             return Optional.empty();
         }
     }
+
+    public RowExpression toRowExpression(Subfield subfield, Type columnType)
+    {
+        List<Subfield.PathElement> path = subfield.getPath();
+        List<Type> types = new ArrayList<>();
+        types.add(columnType);
+        for (int i = 0; i < path.size(); i++) {
+            Type type = types.get(i);
+            if (type instanceof RowType) {
+                types.add(getFieldType((RowType) type, ((Subfield.NestedField) path.get(i)).getName()));
+            }
+            else if (type instanceof ArrayType) {
+                types.add(((ArrayType) type).getElementType());
+            }
+            else if (type instanceof MapType) {
+                types.add(((MapType) type).getValueType());
+            }
+            else {
+                verify(false, "Unexpected type: " + type);
+            }
+        }
+
+        return toRowExpression(subfield, types);
```

Comment:
Return an immutableList.  You can use Immutablelist.builder() to create it in the first place instead of creating an ArrayList and copying. (see https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines)

Reply:
(y)